components/sup/src/http_gateway.rs

```diff
@@ -223,34 +225,73 @@ impl Server {
                     .configure(routes)
             }).workers(thread_count);
 
-            let bind = if cert_file.is_some() && key_file.is_some() {
+            let mut tls_status: Option<ServerStartup> = None;
+            let bind;
+
+            if cert_file.is_some() && key_file.is_some() {
                 let mut config = ServerConfig::new(NoClientAuth::new());
+
+                // The multiple unwraps here are safe because we've already verified that cert_file
+                // and key_file are both Some() and that they point to a file that exists on disk
+                // (otherwise they'd be None)
                 let cert_file = &mut BufReader::new(File::open(cert_file.unwrap()).unwrap());
                 let key_file = &mut BufReader::new(File::open(key_file.unwrap()).unwrap());
-                let cert_chain = certs(cert_file).unwrap();
-                let mut keys = rsa_private_keys(key_file).unwrap();
-                config.set_single_cert(cert_chain, keys.remove(0)).unwrap();
 
-                server.bind_rustls(listen_addr.to_string(), config)
+                let cert_chain = match certs(cert_file) {
+                    Ok(c) => c,
+                    Err(_) => {
+                        tls_status = Some(ServerStartup::InvalidCertFile);
+                        vec![]
+                    }
+                };
+
+                // If we have an invalid cert file, don't even bother checking the key file. The
+                // whole thing is going to error anyway.
+                if tls_status.is_none() {
+                    if let Ok(mut keys) = rsa_private_keys(key_file) {
+                        config.set_single_cert(cert_chain, keys.remove(0)).unwrap();
+                    } else {
+                        tls_status = Some(ServerStartup::InvalidKeyFile);
+                    }
+                }
+
+                // bind has to be initialized to something, or the match statement below is going
+                // to fail. If we've gotten a TLS error already, try to do a regular bind, just so
+                // that that variable has _something_ in it. We'll never use it though, because the
+                // first match arm below will be the path we take.
```

Comment:
That suggests that `bind` should be an `Option`... if you have a TLS error, set it to `None`; if you get to the end and it's still `None`, you know bad stuff happened.

On the other hand, a lot of this logic could be extracted into a separate function, and we could just do an early return with whatever error we happen to find.

Reply:
I agree that this could use some refactoring. The way it reads now feels like it's awkward because it's not following rust idioms. I'd try breaking it up like @christophermaier suggests. Especially try to avoid variables like `tls_status` which are keeping track of the flow of the execution in favor of using helper functions and structural flow control. That will make the behavior of the code more apparent from the block structure rather than having to trace through it to understand.

Additionally, these ideas may help move it toward a more natural and idiomatic approach:
- If state of multiple variables is related, tie them together (e.g. maybe `cert_file` and `key_file` should be `Option<(PathBuf, PathBuf)>` instead of 2 independent options)
- `Option` is about something that may be present, `tls_status` seems like more like a `Result`.
- `unwrap` is usually a sign that there's a more elegant `match` or `if let` approach. Or perhaps it means that the caller should be passing in the non-optional value, or that the `Option<T>`s can be shadowed with safely extracted `T`s:
  ```rust
  fn foo(x: Option<i32>) {
      if let Some(x) = x {
          … // use x without needing to unwrap
      }
      // or
      let x = x.unwrap_or_default();
      … // use x without needing to unwrap   
  }
  ```
- For `mut` variables that aren't keeping track of a process we care about (like a counter or a value built up gradually), try making them (potentially uninitialized) immutable `let`s and take advantage of the compiler's analysis that they're assigned exactly once.
- Rather than assigning the same variable in separate arms of a match or condition:
  ```rust
  if cond {
      foo = bar;
  } else {
      foo = baz;
  }
  ```
  assign once with the value of the match or condition:
  ```rust
  foo = if cond {
      bar
  } else {
      baz
  }
  ```

Reply:
Yeah, totally fair.  The code felt bad when I was writing it, but I didn't have any better ideas at the time.  I've refactored this a fair amount now, and I think the results are better.